torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,9 +1160,60 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR)
+      << "No PGNCCL's watchdog heartbeat detected, so we are dumping debug info: "
+      << dump_nccl_trace();
+}
+
+void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+  // Logging with `FATAL`, after errMsg printed, it calls `std::abort()`
+  // to terminate the program execution.
+  LOG(FATAL) << errMsg;
+}
+
+void ProcessGroupNCCL::heartbeatMonitor() {
+  uint64_t heartBeatCounter = 0ULL;
+  while (true) {
+    auto heartbeat = heartbeat_;
+    if (heartbeat > heartBeatCounter) {
+      heartBeatCounter = heartbeat;
+      // Sleep for enough time so that heartBeat will at least increase.
+      std::this_thread::sleep_for(std::chrono::milliseconds(
+          kWatchdogThreadSleepMillis * heartbeatTimeoutInSec_));
+    } else {
+      // No heartbeat increase detected and timeout.
+      break;
+    }
+  }
+
+  // In the timeout case and we will first dump the flight recorder
+  // to std::out. Down the road, if we have more complicated or blocking
+  // operations, we might need to use a side thread to do it.
+  if (parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0) > 0) {
+    // TODO: Find the right and proper way to dump the debug info.
+    // We cannot print out debugging info directly.
+    dump_debugging_info();
+  }
+
+  // Create a error message reported from MonitorThread, so
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ", rank_, "] NCCL monitor thread terminated with timeout: ");
```

Comment:
is this incomplete? 

- it looks like you put a `:` and forgot to put the details after it
- it would help to explain more about the timeout.  in this case, it's actually the MonitorThread that timed out not the collective.  (or more precisely, we don't know what caused the MonitorThread to time out).  Could write that this is a system error and please file a bug to pytorch.